### PR TITLE
MLKEM FIPS 203 final

### PR DESCRIPTION
**Pull request by rod-chapman**
This PR updates the specification of ML-KEM to meet the final FIPS 203 specification, issued by NIST on 13th August 2024.

Updated KATs from https://github.com/post-quantum-cryptography/KAT/tree/main/MLKEM
run fine with these changes.

### CODE REVIEW

**Review by marsella — APPROVED** [review]
Thank you for updating the code! I confirmed that the KATs run as-is.

This is now out of alignment with the `README` in this directory which points to the initial draft and corresponding KATs, and the PDF of the IPD. I can pop that change into a separate issue for us to handle internally, but you're also welcome to change the links and delete the PDF.

**Comment by marsella** [comment]
I added the two issues that I think this addresses in the sidebar, please remove them if you think they're not fully covered by this PR.

File: Primitive/Asymmetric/Cipher/ML-KEM/specification.cry [hunk 1]
```diff
@@ -500,8 +500,8 @@ property CorrectnessEncodeDecode' f = Decode'`{12}(Encode'`{12} f) == f
 
 K_PKE_KeyGen: ([32]Byte) -> ([384*k+32]Byte, [384*k]Byte)
 K_PKE_KeyGen(d) = (ekPKE, dkPKE) where
-  (ρ,σ) = G(d)
-  A_hat = [[SampleNTT (XOF(ρ,i,j)) | i <- [0 .. k-1]] | j <- [0 .. k-1]] : [k][k]Z_q_256
+  (ρ,σ) = G(d # [`(k)])
+  A_hat = [[SampleNTT (XOF(ρ # [j] # [i])) | j <- [0 .. k-1]] | i <- [0 .. k-1]] : [k][k]Z_q_256
```
[the remark this answers is not in the document]
Comment (rod-chapman) [reply Primitive/Asymmetric/Cipher/ML-KEM/specification.cry:504]: As you wish

File: Primitive/Asymmetric/Cipher/ML-KEM/specification.cry [hunk 2]
```diff
@@ -516,14 +516,14 @@ K_PKE_Encrypt : ([384*k+32]Byte, [32]Byte, [32]Byte) -> [32*(d_u*k+d_v)]Byte
 K_PKE_Encrypt(ekPKE, m, r) = c where
   t_hat = Decode`{12} (ekPKE@@[0 .. 384*k - 1])
   rho = ekPKE@@[384*k .. 384*k + 32 - 1]
-  A_hat = [[SampleNTT (XOF(rho,i,j)) | i <- [0 .. k-1]] | j <- [0 .. k-1]] : [k][k]Z_q_256
-  rvec = [SamplePolyCBD`{eta_1}(PRF(r,N)) | N <- [0 .. k-1]] : [k]Z_q_256
+  A_hat = [[SampleNTT (XOF(rho # [j] # [i])) | j <- [0 .. k-1]] | i <- [0 .. k-1]] : [k][k]Z_q_256
```
[the remark this answers is not in the document]
Comment (rod-chapman) [reply Primitive/Asymmetric/Cipher/ML-KEM/specification.cry:519]: Yes... this fixes #71

File: Primitive/Asymmetric/Cipher/ML-KEM/specification.cry [hunk 3]
```diff
@@ -500,8 +500,8 @@ property CorrectnessEncodeDecode' f = Decode'`{12}(Encode'`{12} f) == f
 
 K_PKE_KeyGen: ([32]Byte) -> ([384*k+32]Byte, [384*k]Byte)
 K_PKE_KeyGen(d) = (ekPKE, dkPKE) where
-  (ρ,σ) = G(d)
-  A_hat = [[SampleNTT (XOF(ρ,i,j)) | i <- [0 .. k-1]] | j <- [0 .. k-1]] : [k][k]Z_q_256
+  (ρ,σ) = G(d # [`(k)])
```
[the remark this answers is not in the document]
Comment (rod-chapman) [reply Primitive/Asymmetric/Cipher/ML-KEM/specification.cry:503]: Yes... this is the big technical difference between IPD and FINAL.

**Comment by rod-chapman** [comment]
Looks good to me.

**Comment by marsella** [comment]
I added #135 to follow up on the other issues I noted here. Thanks for the update!

**Pull request merged**